pilopt/src/lib.rs

```diff
@@ -620,3 +640,61 @@ fn remove_duplicate_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
         .collect();
     pil_file.remove_identities(&to_remove);
 }
+
+/// Identifies witness columns that are directly constrained to be equal to other witness columns
+/// through polynomial identities of the form "x = y".
+fn equal_constrained<T: FieldElement>(
+    expression: &AlgebraicExpression<T>,
+) -> Option<(PolyID, (String, PolyID))> {
+    match expression {
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
+            left,
+            op: AlgebraicBinaryOperator::Sub,
+            right,
+        }) => match (left.as_ref(), right.as_ref()) {
+            (
+                AlgebraicExpression::Reference(left_ref),
+                AlgebraicExpression::Reference(right_ref),
+            ) => {
+                if left_ref.is_witness()
+                    && !left_ref.next
+                    && right_ref.is_witness()
+                    && !right_ref.next
+                {
+                    Some((
+                        left_ref.poly_id,
+                        (right_ref.name.clone(), right_ref.poly_id),
+                    ))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+fn remove_equal_constrained_witness_columns<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let substitutions = pil_file
+        .identities
+        .iter()
+        .filter_map(|id| {
+            if let Identity::Polynomial(PolynomialIdentity { expression: e, .. }) = id {
+                equal_constrained(e)
+            } else {
+                None
+            }
+        })
+        .collect::<HashMap<PolyID, _>>();
+
+    pil_file.post_visit_expressions_in_identities_mut(&mut |e: &mut AlgebraicExpression<_>| {
```

Comment:
I'm wondering if this is enough. I think we also need to visit references in definitions. Can you add a test where you have `x = y` and use `provide_value` on both in a prover function?

Reply:
Ah is that why you commented out `set_hint` below? :)

Reply:
Yes, but also because `remove_unreferenced_definitions` did not seem to take into account the variables inside `definitions`. I added a small case inside this function and set_hint now seems to works perfectly, but I'm sure that the way in which I identify these variables/columns is not 100% correct/sufficiently accurate.